lib/ims/lti/models/messages/Message.rb

```diff
@@ -0,0 +1,57 @@
+require 'active_model'
+
+module Ims::Lti::Models::Messages
+  # Abstract base class for all messages.
+  #
+  # @attr [Hash] extensions Message parameters not defined in the specification.
+  # @attr [Hash] custom Parameters specified by the tool provider.
+  class Message
+    include ActiveModel::Model
+    include ActiveModel::Serialization
+    include ActiveModel::Validations
+
+    attr_accessor :extensions, :custom
+
+    def initialize(params = {})
+      singleton_class.class_eval do
+        attr_accessor *(required_parameters + recommended_parameters + optional_parameters)
+        validates *required_parameters, presence: true
+      end
+      super
+    end
+
+    class << self
+      def required_parameters
+        raise 'Override "required_parameters" in child class'
+      end
+
+      def recommended_parameters
+        raise 'Override "recommended_parameters" in child class'
```

Comment:
This and optiona_parameters should return a default value since they aren't technically required.